### PR TITLE
Switch to Akka HTTP based AWS connector

### PR DESCRIPTION
**Pull request by christianuhlcc**
- Removes AWS SDK
- Adds Alpakka as new Client
- To perform a clean change we refactored the DynamoDBJournal so that the Helper can be replaced in isolation.
- Introduces settings for Backoff and Retries instead of magic numbers.
- Refactored Describe from Type Class to pattern matching for better testablility and simpler code
- Removed no longer used AWS Client Settings
- Removed Documentation of Threading Issue from Readme
- Bump Version to 1.1.0
- Use default AWS credential provider chain instead of config values

This fixes issue #3

### CODE REVIEW

**Review by hseeberger — COMMENTED** [review]
Please upgrade to sbt 0.13.13.

I have verified the PR locally, no idea why it has failed on Travis CI.

**Comment by hseeberger** [comment]
@ktoso do you think you can find time for a review?

**Comment by ktoso** [comment]
I'll try tomorrow, somewhat chaotic weeks for me (moving)

**Comment by christianuhlcc** [comment]
I updated sbt to 13.13 and Scala to 2.11.11 / 2.12.2 since the world has moved on by now
Hope your move went well @ktoso !

File: README.md [hunk 1]
```diff
@@ -36,16 +36,18 @@ Configuration
...
 ~~~
-akka.persistence.journal.plugin = "my-dynamodb-journal"
-
-my-dynamodb-journal = ${dynamodb-journal} # include the default settings
-my-dynamodb-journal {                     # and add some overrides
-    journal-table =  <the name of the table to be used>
-    journal-name =  <prefix to be used for all keys stored by this plugin>
-    aws-access-key-id =  <your key>
-    aws-secret-access-key =  <your secret>
-    endpoint =  "https://dynamodb.us-east-1.amazonaws.com" # or where your deployment is
```
Comment (patriknw) [inline README.md:47]: why was this removed, it showed how to configure a custom journal, when using more than one plugin

File: README.md [hunk 2]
```diff
@@ -73,17 +78,6 @@ In the first case a recovery will only ever see all of the events or none of the
...
-This plugin uses the AWS Java SDK which means that the number of requests that can be made concurrently is limited by the number of connections to DynamoDB and by the number of threads in the thread-pool that is used by the AWS HTTP client. The default setting is 50 connections which for a deployment that is used from the same EC2 region allows roughly 5000 requests per second (where every persisted event batch is roughly one request). If a single ActorSystem needs to persist more than this number of events per second then you may want to tune the parameter
-
-~~~
-my-dynamodb-journal.aws-client-config.max-connections = <your value here>
-~~~
-
-Changing this number changes both the number of concurrent connections and the used thread-pool size.
-
 Compatibility with pre-1.0 versions
 -----------------------------------
```
Comment (patriknw) [inline README.md:83]: perhaps add a section with migration instructions from 1.0 to 1.1

File: README.md [hunk 3]
```diff
@@ -73,17 +78,6 @@ In the first case a recovery will only ever see all of the events or none of the
 
 In the second case each event is treated in isolation and may or may not be replayed depending on whether it was persisted successfully or not.
 
-Performance Considerations
---------------------------
-
-This plugin uses the AWS Java SDK which means that the number of requests that can be made concurrently is limited by the number of connections to DynamoDB and by the number of threads in the thread-pool that is used by the AWS HTTP client. The default setting is 50 connections which for a deployment that is used from the same EC2 region allows roughly 5000 requests per second (where every persisted event batch is roughly one request). If a single ActorSystem needs to persist more than this number of events per second then you may want to tune the parameter
```
Comment (patriknw) [inline README.md:79]: similar considerations with this implementation?

File: src/main/resources/reference.conf [hunk 4]
```diff
@@ -120,4 +121,13 @@ dynamodb-journal {
       parallelism-max = 8
     }
   }
+
+  akka.stream.alpakka.dynamodb {
```
Comment (patriknw) [inline src/main/resources/reference.conf:125]: that is probably a reference conf section in alpakka? you can't override that from another reference.conf

File: src/main/scala/akka/persistence/dynamodb/journal/DynamoDBHelper.scala [hunk 5]
```diff
@@ -1,134 +1,108 @@
...
-  object Describe {
-    implicit object GenericDescribe extends Describe[AmazonWebServiceRequest] {
-      def desc(aws: AmazonWebServiceRequest): String = aws.getClass.getSimpleName
+  def retry[A](backoff: FiniteDuration, retries: AtomicInteger)(f: () => Future[A]): Future[A] = {
+    f().recoverWith {
+      case e: ProvisionedThroughputExceededException =>
+        if (retries.getAndDecrement() > 0) {
+          Future {
+            log.error(s"ProvisionedThroughputExceededException, backing of $backoff ms, retries left ${retries.get()}")
+            Thread.sleep(backoff.toMillis)
```
Comment (patriknw) [inline src/main/scala/akka/persistence/dynamodb/journal/DynamoDBHelper.scala:82]: Thread.sleep should not be needed. Perhaps use http://doc.akka.io/docs/akka/2.4/scala/futures.html#After

**Review by patriknw — COMMENTED** [review]
looking good, but I'm not sure I understand the full impact

**Comment by raboof** [comment]
@christianuhlcc Seems like a neat improvement - would you be interested in looking at the review comments and fixing merge conflicts?

**Comment by christianuhlcc** [comment]
First of all sorry everybody for keeping this around way to long, life got somehow in the way. Especially to @patriknw for basically ignoring your review comments!

Since I didn't really follow the development of this module in the last months, does this PR still make sense in the way it was build @ktoso? If it's still appreciated i could spend the time to deal with the conflicts, prepare answers to the review and maybe also provide a working example in a different repo

**Comment by teroxik** [comment]
I makes more sense to go for the AWS V2 implementation, we can look into the alpakka / akka stream approach or just use the plain java client. @christianuhlcc WDYT?

**Comment by christianuhlcc** [comment]
Sure, stuff has changed by now, guess it makes sense to start over fresh. Unfortunately I haven't worked in the Scala / Akka Ecosystem for quite some time, I'm not capable to do it on my own right now. @teroxik would be willing to spend some time remote-pairing with me to work out the rough plan?

**Comment by teroxik** [comment]
@christianuhlcc I was planning to do the AWS v2 myself, but if you want to be involved, help is always welcome.

**Comment by christianuhlcc** [comment]
I'm probably just in your way @teroxik, so let's close this PR and you can have a go at V2 :)

**Comment by teroxik** [comment]
No worries sorry. Have been busy couple of past weeks. Made a good progress. Mainly have to fix the tests other than that I have a working version of AWS v2.
